parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala

```diff
@@ -414,54 +412,42 @@ base match {
+    body match {
+      case Lambda(args, body) =>
+        def rewrite(tree: Term): Term =
+          tree.tpe.asType match {
+            case '[Rule[_, _]] => expand(tree, wrapped).asInstanceOf[Term]
+            case _ =>
+              tree match {
+                case Block(statements, res) => block(statements, rewrite(res))
+                case x                      => '{ $parser.__push(${ x.asExpr }) }.asTerm
+              }
+          }
+        // do a beta reduction, using the parameter definitions as stubs for variables
+        // that hold values popped from the stack
+        block(popToVals(args), rewrite(body)).asExprOf[Boolean]
     }
-    def ruleTraceNonTerminalKey = reify(RuleTrace.Action).tree
-
-    def renderInner(wrapped: Boolean): Tree = {
-      val argTypes = actionType dropRight 1
-
-      def popToVals(valNames: List[TermName]): List[Tree] =
-        (valNames zip argTypes).map { case (n, t) => q"val $n = valueStack.pop().asInstanceOf[$t]" }.reverse
+  }
 
-      def actionBody(tree: Tree): Tree =
-        tree match {
-          case Block(statements, res) => block(statements, actionBody(res))
+  private case class Action(body: Term, ts: List[TypeTree]) extends DefaultNonTerminalOpTree {
+    def ruleTraceNonTerminalKey = '{ RuleTrace.Action }
 
-          case x @ (Ident(_) | Select(_, _)) =>
-            val valNames = argTypes.indices.map(i => TermName("value" + i)).toList
-            val args     = valNames map Ident.apply
-            block(popToVals(valNames), q"__push($x(..$args))")
+    def renderInner(start: Expr[Int], wrapped: Boolean): Expr[Boolean] = expandLambda(body, wrapped)
+  }
 
-          case q"(..$args => $body)" =>
-            def rewrite(tree: Tree): Tree =
-              tree match {
-                case Block(statements, res)                                  => block(statements, rewrite(res))
-                case x if isSubClass(actionType.last, "org.parboiled2.Rule") => expand(x, wrapped)
-                case x                                                       => q"__push($x)"
-              }
-            block(popToVals(args.map(_.name)), rewrite(body))
-        }
+  case class RunAction(body: Expr[_]) extends DefaultNonTerminalOpTree {
+    def ruleTraceNonTerminalKey = '{ RuleTrace.Run }
 
-      actionBody(c.untypecheck(actionTree))
-    }*/
+    def renderInner(start: Expr[Int], wrapped: Boolean): Expr[Boolean] =
+      body.asTerm.tpe.asType match {
+        case '[Rule[_, _]]                => expand(body, wrapped)
+        case '[t1 => r]                   => expandLambda(body.asTerm, wrapped)
+        case '[(t1, t2) => r]             => expandLambda(body.asTerm, wrapped)
+        case '[(t1, t2, t3) => r]         => expandLambda(body.asTerm, wrapped)
+        case '[(t1, t2, t3, t4) => r]     => expandLambda(body.asTerm, wrapped)
+        case '[(t1, t2, t3, t4, t5) => r] => expandLambda(body.asTerm, wrapped)
```

Comment:
Is "5 arguments max" an arbitrary limit to do lambda expansion?

Reply:
That's an existing limitation of the `RunResult` typing structure. I don't know the original reason but maybe it's to avoid a similar code explosion as for what's needed to make `~>` work.